src/Rector/ClassMethod/AddGenericReturnTypeToRelationsRector.php

```diff
@@ -0,0 +1,184 @@
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\Generic\GenericClassStringType;
+use PHPStan\Type\ObjectType;
+use Rector\BetterPhpDocParser\PhpDocInfo\PhpDocInfo;
+use Rector\BetterPhpDocParser\ValueObject\Type\FullyQualifiedIdentifierTypeNode;
+use Rector\Core\Rector\AbstractRector;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/** @see \Rector\Laravel\Tests\Rector\ClassMethod\AddGenericReturnTypeToRelationsRector\AddGenericReturnTypeToRelationsRectorTest */
+class AddGenericReturnTypeToRelationsRector extends AbstractRector
+{
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition(
+            'Add generic return type to relations in child of Illuminate\Database\Eloquent\Model',
+            [
+                new CodeSample(
+                    <<<'CODE_SAMPLE'
+use App\Account;
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\Relations\HasMany;
+
+class User extends Model
+{
+    public function accounts(): HasMany
+    {
+        return $this->hasMany(Account::class);
+    }
+}
+CODE_SAMPLE
+
+                    ,
+                    <<<'CODE_SAMPLE'
+use App\Account;
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\Relations\HasMany;
+
+class User extends Model
+{
+    /** @return HasMany<Account> */
+    public function accounts(): HasMany
+    {
+        return $this->hasMany(Account::class);
+    }
+}
+CODE_SAMPLE
+                ),
+            ]
+        );
+    }
+
+    /**
+     * @return array<class-string<Node>>
+     */
+    public function getNodeTypes(): array
+    {
+        return [ClassMethod::class];
+    }
+
+    /**
+     * @param ClassMethod $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        if ($this->shouldSkipNode($node)) {
+            return null;
+        }
+
+        /** @var Node\Stmt[] $methodStatements */
+        $methodStatements = $node->stmts;
+
+        $methodReturnType = $node->getReturnType();
+
+        if ($methodReturnType === null) {
+            return null;
+        }
+
+        $methodReturnTypeName = $this->getName($methodReturnType);
+
+        if ($methodReturnTypeName === null) {
+            return null;
+        }
+
+        if (! $this->isObjectType(
+            $methodReturnType,
+            new ObjectType('Illuminate\Database\Eloquent\Relations\Relation')
+        )) {
+            return null;
+        }
+
+        $docInfo = $this->phpDocInfoFactory->createFromNode($node);
+        if (! $docInfo instanceof PhpDocInfo) {
+            return null;
+        }
+
+        // Return, if already has return type
+        if ($node->getDocComment() !== null && $docInfo->hasByName('return')) {
+            return null;
+        }
+
+        $returnStatement = $this->betterNodeFinder->findFirstInstanceOf($methodStatements, Return_::class);
```

Comment:
You can use `->findFirstInFunctionLikeScoped()` to ensure the return statement is not in the inner function, eg:

```php
        $returnStatement = $this->betterNodeFinder->findFirstInFunctionLikeScoped(
            $node,
            fn (Node $subNode): bool => $subNode instanceof Return_
        );
```

that's check the method stmts, but only if in scoped. By this, the following code will be skipped:

```php
function toBeChecked()
{
   function inner()
   {
         return X;
   }
}
```

Reply:
Thanks. Changed.